File: content/snippets/python/problems/balanced_brackets.py
def is_balanced(s: str) -> bool:
    close_to_open = {")": "(", "]": "[", "}": "{"}
    stack: list[str] = []

    for ch in s:
        if ch in "([{":
            stack.append(ch)
        elif ch in close_to_open:
            if not stack or stack[-1] != close_to_open[ch]:
                return False
            stack.pop()

    return not stack


def main() -> None:
    input_1 = "([{}])(()[]){}"
    input_2 = "([)]"

    print(f"input_1 valid: {str(is_balanced(input_1)).lower()}")
    print(f"input_2 valid: {str(is_balanced(input_2)).lower()}")


if __name__ == "__main__":
    main()

File: content/snippets/typescript/problems/balanced_brackets.ts
import { setImmediate } from "node:timers";

function isBalanced(s: string): boolean {
  const closeToOpen: Record<string, string> = {
    ")": "(",
    "]": "[",
    "}": "{",
  };

  const stack: string[] = [];
  for (const ch of s) {
    if (ch === "(" || ch === "[" || ch === "{") {
      stack.push(ch);
    } else if (ch in closeToOpen) {
      if (stack.length === 0 || stack[stack.length - 1] !== closeToOpen[ch]) {
        return false;
      }
      stack.pop();
    }
  }

  return stack.length === 0;
}

function main(): void {
  const input1 = "([{}])(()[]){}";
  const input2 = "([)]";

  console.log(`input_1 valid: ${isBalanced(input1)}`);
  console.log(`input_2 valid: ${isBalanced(input2)}`);
}

main();
setImmediate(() => {});
